server/src/main/java/org/opensearch/node/PerformanceCollectorService.java

```diff
@@ -0,0 +1,152 @@
+     */
+    public void collectNodePerfStatistics(String nodeId, long timestamp, double memoryUtilizationPercent, double cpuUtilizationPercent) {
+        nodeIdToPerfStats.compute(nodeId, (id, nodePerfStats) -> {
+            if (nodePerfStats == null) {
+                return new NodePerformanceStats(nodeId, timestamp, memoryUtilizationPercent, cpuUtilizationPercent);
+            } else {
+                nodePerfStats.cpuUtilizationPercent = cpuUtilizationPercent;
+                nodePerfStats.memoryUtilizationPercent = memoryUtilizationPercent;
+                nodePerfStats.timestamp = timestamp;
+                return nodePerfStats;
+            }
+        });
+    }
+
+    /**
+     * Get all node statistics which will be used for node stats
+     */
+    public Map<String, NodePerformanceStats> getAllNodeStatistics() {
+        Map<String, NodePerformanceStats> nodeStats = new HashMap<>(nodeIdToPerfStats.size());
+        nodeIdToPerfStats.forEach((nodeId, nodePerfStats) -> { nodeStats.put(nodeId, new NodePerformanceStats(nodePerfStats)); });
+        return nodeStats;
+    }
+
+    /**
+     * Optionally return a {@code NodePerformanceStatistics} for the given nodeid, if
+     * performance stats information exists for the given node. Returns an empty
+     * {@code Optional} if the node was not found.
+     */
+    public Optional<NodePerformanceStats> getNodeStatistics(final String nodeId) {
+        return Optional.ofNullable(nodeIdToPerfStats.get(nodeId)).map(perfStats -> new NodePerformanceStats(perfStats));
+    }
+
+    /**
+     * Returns collected performance statistics of all nodes
+     */
+    public NodesPerformanceStats stats() {
+        return new NodesPerformanceStats(getAllNodeStatistics());
+    }
+
+    /**
+     * Fetch local node performance statistics and add it to store along with the current timestamp
+     */
+    private void collectLocalNodePerformanceStats() {
+        if (nodePerformanceTracker.isReady() && clusterService.state() != null) {
+            collectNodePerfStatistics(
+                clusterService.state().nodes().getLocalNodeId(),
+                System.currentTimeMillis(),
+                nodePerformanceTracker.getMemoryUtilizationPercent(),
+                nodePerformanceTracker.getCpuUtilizationPercent()
+            );
+        }
+    }
+
+    @Override
+    protected void doStart() {
+        /**
+         * Fetch local node performance statistics every second
+         */
+        scheduledFuture = threadPool.scheduleWithFixedDelay(() -> {
+            try {
+                collectLocalNodePerformanceStats();
+            } catch (Exception e) {
+                logger.warn("failure in PerformanceCollectorService", e);
+            }
+        }, new TimeValue(REFRESH_INTERVAL_IN_MILLIS), ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    protected void doStop() {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel();
+        }
+    }
```

Comment:
Should close() also have the same logic?

Reply:
as part of close() we first do stop() and then close() , so this logic won't be needed. Reference - https://github.com/opensearch-project/OpenSearch/blob/main/libs/common/src/main/java/org/opensearch/common/lifecycle/AbstractLifecycleComponent.java#L110